Changes to speed up qvm-ls

qubesuser opened the pull request:
This is the server part of a bunch of changes that get qvm-ls to around 200ms run time.

Warning: I haven't tested them much, and they change fundamental code, so they probably need fixes

The first major change is the introduction of APIs that allow to get all properties of a VM or of Qubes and all properties of all VMs in a single call.

The second major change is storing the libvirt state in qubesd (and changing it in response to events), so that libvirtd calls are not required to get the system state.

Finally, there's some optimization and fixes.

qubesuser committed:
Add admin APIs to get all properties at once, and optimize the code

This allows MUCH faster qvm-ls and qvm-prefs

The code is optimized to be as fast as possible:
- optimized property access
- gather strings in a list and join them only at the end

qubesuser committed:
cache PropertyHolder.property_list and use O(1) property name lookups

qubesuser committed:
rework libvirt connection code

- connect in a loop, starting libvirtd
- reconnect on demand in a loop
- register events on reconnection

qubesuser committed:
store libvirt state and update on events instead of asking libvirtd every time

qubesuser committed:
take startup lock while starting services

This should prevent getting "qrexec not connected" due to trying
to start a service in the middle of VM startup

qubesuser committed:
don't ask for domain state before asking for info

info already contains domain state, so just use it, making a single
libvirtd call

qubesuser committed:
don't replace pdb debugger's SIGINT handler

Otherwise qubesd will die if you try to break in the debugger

qubesuser committed:
cache isinstance(default, collections.Callable)

qubesuser committed:
don't access netvm if it's None in visible_gateway/netmask

Causes an unnecessary exception

qubesuser committed:
add GetAllData API to get data of all VMs

Even faster qvm-ls!

qubesuser committed:
change default pool code to be fast

currently it takes 100ms+ to determine the default pool every time,
including subprocess spawning (!), which is unacceptable since
finding out the default value of properties should be instantaneous

instead of checking every thin pool to see if the root device is in
it, find and cache the name of the root device thin pool and see if
there is a configured pool with that name

marmarek commented on qubes/api/admin.py:
```diff
@@ -119,11 +140,10 @@ def vm_list(self):
+        strs = []
+        for vm in sorted(domains):
+            self._vm_line_strs(strs, vm)
+        return ''.join(strs)
```
Is `str.append` really faster than `str.format`?

qubesuser replied:
The idea is to avoid doing any kind of string processing until the final join call, which should be the fastest since it should just allocate a single string in join and copy them all in, avoiding any intermediate allocations, partial concatenations, format string parsing, etc.

It's slightly uglier, but not that much, it looks like writing to an output stream (which would be even faster, but not really worth the effort).

qubesuser committed:
create "lvm" pool using rootfs thin pool instead of hardcoding qubes_dom0-pool00

marmarek commented on qubes/api/admin.py:
```diff
@@ -145,6 +145,33 @@ def vm_list(self):
             self._vm_line_strs(strs, vm)
         return ''.join(strs)
 
+    @qubes.api.method('admin.vm.GetAllData', no_payload=True,
```
Sorry, this is too much. Even if we agree for `property.GetAll`, getting all info of all the VMs in one call is beyond what we want in Admin API.

marmarek commented on qubes/app.py:
```diff
@@ -553,6 +553,38 @@ def get_new_unused_dispid(self):
+                pass
+        return cls._name
+
 def _default_pool(app):
```
`@functools.lru_cache` instead?

marmarek reviewed (changes requested):
Some of those commits are good as-is, some requires changes. We'll think about `property.GetAll`, so if you think other changes are important improvements anyway, splitting it into separate PR may be a good idea.

qubesuser commented on qubes/vm/qubesvm.py:
```diff
@@ -1026,22 +1026,24 @@ def run_service(self, service, source=None, user=None,
+            elif not self.is_running():
+                if not autostart:
+                    raise qubes.exc.QubesVMNotRunningError(self)
+                yield from self.start(start_guid=gui)
```
Changed, this commit is quite unrelated to the rest anyway.

qubesuser commented on qubes/vm/qubesvm.py:
```diff
@@ -695,6 +683,49 @@ def __xml__(self):
+                    self._stubdom_xid = int(stubdom_xid_str)
+
+        if self.is_halted() and not was_halted:
+            self.fire_event('domain-shutdown')
```
This prevents the "double shutdown event" as well. It think just taking the startup lock while firing the event could fix the synchronization issue?

marmarek replied:
It will not prevent handling domain-shutdown after VM was started again, namely this sequence:
1. VM shutdown
2. vm.start()
3. vm.fire_event('domain-shutdown')

You don't have any guarantee about order of input processing (request on qubesd.sock, event on libvirt socket). And additionally, you can't take lock directly in libvirt event handler, because it isn't coroutine.

qubesuser replied:
So I think this needs to be thought about and done more carefully: probably the simplest way is having non-coroutines schedule coroutines, and then pretty much serializing all methods that change the VM in any way, possibly including libvirt state updates, by taking the startup_lock (which should just be renamed to "lock") and updating libvirt state under lock if necessary.

Could also consider deleting the libvirt domain for VMs that are shutdown to prevent external tools starting it while we finalize storage and other such robustness measures against raw use of the libvirt/xen APIs.

I think neither the current code nor this patch does this correctly, and some more work is needed.

qubesuser commented on qubes/vm/qubesvm.py:
```diff
@@ -1446,60 +1495,32 @@ def get_power_state(self):
-            raise
-
-        assert False
+        state = self._libvirt_state
```
The idea is to never call libvirtd for any read-only admin API call except for CPU/memory stats (and internal inspection).

It should theoretically not be fragile because it updates the state on every connection to libvirtd, on every lifecyle event received, and if the libvirt connection dies the close callback should be called, which triggers a reconnect, which triggers a state update.

Of course it's possible that the implementation is buggy

Without this doing qvm-ls requires at least one call to libvirtd per VM, which is probably going to take hundreds of milliseconds, and also any random "is_running" and similar call within qubes is going to cause I/O, which can also cause pervasive slowdown.

qubesuser commented on qubes/app.py:
```diff
@@ -90,27 +90,45 @@ def __getattr__(self, attrname):
+                self._conn.registerCloseCallback(self._close_callback, None)
+                break
+            except libvirt.libvirtError:
+                subprocess.call(["systemctl", "start", "libvirtd"])
```
I removed it, although not sure which way is the best. The problem of not doing this is that qubesd will just freeze forever while trying to reconnect if libvirtd is stopped, which also doesn't seem great.

marmarek replied:
Maybe add some timeout (5s?) and throw exception?

qubesuser committed:
remove unused netid code

it's unused and has a netid property with name different than key
that would cause issues in the next commit

marmarek commented:
The property.GetAll idea has been implemented in #298. Few individual commits may be worth including too, but the IMO most important part already is.

ben-grande commented:
> This is the server part of a bunch of changes that get qvm-ls to around 200ms run time.

We don't have:

- the command used to time this PR at that tim
-  how many qubes the user had
- how much time it took before this PR at that time

So I don't know how much we can compare, anyway, a simple comparison with the current state below.

Amount of qubes:
```
% qvm-ls --raw-list --no-spinner | wc -l
44
```

Package version:
```
% rpm -qf /usr/lib/python3.13/site-packages/qubesadmin
python3-qubesadmin-4.4.0-1.37.fc41.noarch

% rpm -qf /usr/lib/python3.13/site-packages/qubes
qubes-core-dom0-4.4.0-1.45.fc41.noarch
```

Test:
```
% i=0; while test $i -le 5; do time qvm-ls >/dev/null; sleep 0.5; i=$((i+1)); done
```

With Qubes R4.4
```
0.10s user 0.03s system 59% cpu 0.222 total
0.18s user 0.04s system 65% cpu 0.334 total
0.13s user 0.04s system 58% cpu 0.293 total
0.12s user 0.05s system 63% cpu 0.262 total
0.10s user 0.04s system 61% cpu 0.229 total
0.10s user 0.03s system 60% cpu 0.214 total

# --no-spinner
0.26s user 0.07s system 70% cpu 0.466 total
0.13s user 0.05s system 46% cpu 0.389 total
0.12s user 0.04s system 66% cpu 0.244 total
0.10s user 0.03s system 53% cpu 0.251 total
0.09s user 0.04s system 59% cpu 0.218 total
0.25s user 0.08s system 61% cpu 0.529 total
```

With https://github.com/QubesOS/qubes-core-admin/pull/819:
```
0.10s user 0.04s system 70% cpu 0.198 total
0.10s user 0.03s system 71% cpu 0.181 total
0.11s user 0.04s system 54% cpu 0.262 total
0.09s user 0.03s system 70% cpu 0.175 total
0.11s user 0.03s system 60% cpu 0.225 total
0.10s user 0.04s system 53% cpu 0.269 total

# --no-spinner
0.11s user 0.03s system 52% cpu 0.271 total
0.23s user 0.07s system 82% cpu 0.364 total
0.10s user 0.03s system 54% cpu 0.248 total
0.14s user 0.05s system 59% cpu 0.318 total
0.10s user 0.04s system 54% cpu 0.257 total
0.10s user 0.03s system 69% cpu 0.184 total
```

I think there still is space for improvement, if someone wants to take it, I'd be happy.

Also see benchmark results on https://github.com/QubesOS/qubes-core-admin-client/pull/36#issuecomment-5114575767.
